src/common/providers/identity.ts

```diff
@@ -34,21 +58,263 @@ export type UserRecord = firebase.auth.UserRecord;
+  aud: string;
+  exp: number;
+  iat: number;
+  iss: string;
+  sub: string;
+  event_id: string;
+  event_type: string;
+  ip_address: string;
+  user_agent?: string;
+  locale?: string;
+  sign_in_method?: string;
+  user_record?: DecodedJwtUserRecord;
+  tenant_id?: string;
+  raw_user_info?: string;
+  sign_in_attributes?: {
+    [key: string]: any;
+  };
+  oauth_id_token?: string;
+  oauth_access_token?: string;
+  oauth_refresh_token?: string;
+  oauth_token_secret?: string;
+  oauth_expires_in?: number;
+  [key: string]: any;
+}
+
 /**
  * Helper class to create the user metadata in a UserRecord object
  */
-export class UserRecordMetadata implements firebase.auth.UserMetadata {
+export class UserRecordMetadata implements UserMetadata {
   constructor(public creationTime: string, public lastSignInTime: string) {}
 
   /** Returns a plain JavaScript object with the properties of UserRecordMetadata. */
-  toJSON() {
+  toJSON(): object {
     return {
       creationTime: this.creationTime,
       lastSignInTime: this.lastSignInTime,
     };
   }
 }
 
+/**
+ * Helper class to create the user info in a UserRecord object
+ */
+export class UserRecordInfo implements UserInfo {
+  constructor(
+    public uid: string,
+    public displayName: string,
+    public email: string,
+    public photoURL: string,
+    public providerId: string,
+    public phoneNumber: string
+  ) {}
+
+  toJSON(): object {
+    return {
+      uid: this.uid,
+      displayName: this.displayName,
+      email: this.email,
+      photoURL: this.photoURL,
+      providerId: this.providerId,
+      phoneNumber: this.phoneNumber,
+    };
+  }
+}
+
+/**
+ * Helper class to create the user MultiFactorInfo in a UserRecord object
+ */
+export class UserRecordMultiFactorInfo
+  implements Pick<MultiFactorInfo, keyof MultiFactorInfo> {
+  constructor(
+    public uid: string,
+    public factorId: string,
+    public displayName?: string,
+    public enrollmentTime?: string
+  ) {}
+
+  toJSON(): object {
+    return {
+      uid: this.uid,
+      factorId: this.factorId,
+      displayName: this.displayName || null,
+      enrollmentTime: this.enrollmentTime || null,
+    };
+  }
+}
+
+/**
+ * Helper class to create the user PhoneMultiFactorInfo in a UserRecord object
+ */
+export class UserRecordPhoneMultiFactorInfo
+  implements Pick<PhoneMultiFactorInfo, keyof PhoneMultiFactorInfo> {
+  constructor(
+    public uid: string,
+    public factorId: string,
+    public phoneNumber: string,
+    public displayName?: string,
+    public enrollmentTime?: string
+  ) {}
+
+  toJSON(): object {
+    return {
+      uid: this.uid,
+      factorId: this.factorId,
+      phoneNumber: this.phoneNumber,
+      displayName: this.displayName || null,
+      enrollmentTime: this.enrollmentTime || null,
+    };
+  }
+}
+
+/**
+ * Helper class to create the user MultiFactorSettings in a UserRecord object
+ */
+export class UserRecordMultiFactorSettings implements MultiFactorSettings {
```

Comment:
You've got a lot of classes implementing interfaces and then reexporting the toJSON that would have worked for those interfaces to begin with. What value are we getting here over just using interfaces?

Reply:
I ended up refactoring all the classes into interfaces in the latest commit.